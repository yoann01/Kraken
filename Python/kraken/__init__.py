"""Kraken Framework."""

import logging
import os

__all__ = ['core', 'helpers', 'plugins', 'ui']


krakenPath = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..'))
if os.environ.get('KRAKEN_PATH', None) is None:
    os.environ['KRAKEN_PATH'] = krakenPath

krakenExtsPath = os.path.join(krakenPath, 'Exts')
if 'FABRIC_EXTS_PATH' in os.environ:
    if krakenExtsPath not in os.environ['FABRIC_EXTS_PATH']:
        os.environ['FABRIC_EXTS_PATH'] = krakenExtsPath + os.pathsep + os.environ['FABRIC_EXTS_PATH']

krakenPresetsPath = os.path.join(krakenPath, 'Presets', 'DFG')
if 'FABRIC_DFG_PATH' in os.environ:
    if krakenPresetsPath not in os.environ['FABRIC_DFG_PATH']:
        os.environ['FABRIC_DFG_PATH'] = krakenPresetsPath + os.pathsep + os.environ['FABRIC_DFG_PATH']


logging.basicConfig(format='[KRAKEN] %(levelname)s: %(message)s', level=logging.INFO)

# Custom inform level for use with UI label getting added to the status bar.
logging.INFORM = 25
logging.addLevelName(logging.INFORM, 'INFORM')
logging.Logger.inform = lambda inst, msg, *args, **kwargs: inst.log(logging.INFORM, msg, *args, **kwargs)


def release():
    print "RELEASE THE KRAKEN!!!!"
