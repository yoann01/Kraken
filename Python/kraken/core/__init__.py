"""Kraken Core."""

VERSION_MAJOR = 1
VERSION_MINOR = 2
VERSION_BUILD = 6
VERSION_SUFFIX = ""


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Returns:
        str: Current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
