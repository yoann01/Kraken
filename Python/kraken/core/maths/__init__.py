"""Kraken - math module."""

from math_object import MathObject
from vec2 import Vec2
from vec3 import Vec3
from vec4 import Vec4
from quat import Quat
from euler import Euler
from xfo import Xfo
from mat33 import Mat33
from mat44 import Mat44
from rotation_order import RotationOrder
from color import Color
from constants import DEG_TO_RAD, RAD_TO_DEG, PI


def Math_radToDeg(val):
    """Converts radians to degrees.

    Args:
        val (float): value to convert to degrees.

    Returns:
        float: Value in degrees.

    """

    return val * RAD_TO_DEG

def Math_degToRad(val):
    """Converts degrees to radians.

    Args:
        val (float): value to convert to radians.

    Returns:
        float: Value in radians.

    """

    return val * DEG_TO_RAD


def decodeValue(jsonData):
    """Returns a constructed math value based on the provided json data.

    Args:
        jsondata (dict): The JSON data to use to decode into a Math value.

    Returns:
        object: The constructed math value

    """

    if type(jsonData) is not dict:
        return jsonData

    if '__mathObjectClass__' not in jsonData:
        raise Exception("Invalid JSON data for constructing value:" + str(jsonData))

    if jsonData['__mathObjectClass__'] == 'Vec2':
        val = Vec2()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Vec3':
        val = Vec3()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Vec4':
        val = Vec4()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Euler':
        val = Euler()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Quat':
        val = Quat()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Xfo':
        val = Xfo()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Mat33':
        val = Mat33()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Mat44':
        val = Mat44()
        val.jsonDecode(jsonData, decodeValue)
    else:
        raise Exception("Unsupported Math type:" + jsonData['__mathObjectClass__'])

    return val
