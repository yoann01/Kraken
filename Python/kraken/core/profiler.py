"""Kraken - core.profiler module.

Classes:
Profiler - Profiler Object.

"""

import time
import operator


class _ProfilerItem(object):

    def __init__(self, label):
        super(_ProfilerItem, self).__init__()

        t = time.time()
        self.label = label
        self.start = t
        self.end = t
        self.children = []


    def addChild(self, item):
        self.children.append(item)


    def endProfiling(self):
        self.end = time.time()


class Profiler(object):
    """Kraken profiler object for debugging performance issues."""

    __instance = None


    def __init__(self):
        super(Profiler, self).__init__()
        self.reset()


    def reset(self):
        """Resets the profiler for generating a new report"""

        self.__roots = []
        self.__stack = []


    def push(self, label):

        """Adds a new child to the profiling tree and activates it.

        Args:
            label (str): The label of the next level of the stack.

        """

        item = _ProfilerItem(label)
        if len(self.__stack) == 0:
            self.__roots.append(item)
        else:
            self.__stack[-1].addChild(item)

        self.__stack.append(item)


    def pop(self):
        """Deactivates the current item in the tree and returns the profiler to
        the parent item"""

        if len(self.__stack) == 0:
            raise Exception("""Unable to close bracket. Pop has been called more """ +
                            """times than push.""")

        self.__stack[-1].endProfiling()
        self.__stack.pop()


    def generateReport(self, listFunctionTotals=False):
        """Returns a report string containing all the data gathered turing
        profiling.

        Args:
            listFunctionTotals (bool): list information relating to the total time spent in each function.

        Returns:
            str: The profiler report.

        """

        if len(self.__stack) != 0:
            raise Exception("""Profiler brackets not closed properly. """ +
                            """Pop must be called for every call to push. Pop """ +
                            """needs to be called another """ +
                            str(len(self.__stack)) + """ times""")

        report = []
        report.append("--callstack--")

        functions = {}

        def reportItem(item, indent):
            duration = item.end - item.start
            report.append(indent + item.label + ' duration: ' + str(duration))
            if item.label not in functions:
                functions[item.label] = duration
            else:
                functions[item.label] += duration

            for childItem in item.children:
                reportItem(childItem, indent + '  ')

        for rootItem in self.__roots:
            reportItem(rootItem, '  ')

        if listFunctionTotals:
            report.append("--functions--")

            def reverse_numeric(x, y):
                return y < x

            sorted_fns = sorted(functions.items(), key=operator.itemgetter(1),
                                cmp=reverse_numeric)

            for fn_tuple in sorted_fns:
                report.append(str(fn_tuple[1]) + ': ' + fn_tuple[0])

        return '\n'.join(report)


    @classmethod
    def getInstance(cls):
        """This class method returns the singleton instance for the Profiler

        Returns:
            object: The singleton profiler instance.

        """

        if cls.__instance is None:
            cls.__instance = Profiler()

        return cls.__instance
