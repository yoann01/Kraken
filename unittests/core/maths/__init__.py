
import unittest

import test_vec2
import test_vec3
import test_vec4
import test_color
import test_euler
import test_mat33
import test_mat44
import test_quat
import test_rotation_order
import test_xfo

loadVec2Suite = test_vec2.suite()
loadVec3Suite = test_vec3.suite()
loadVec4Suite = test_vec4.suite()
loadColorSuite = test_color.suite()
loadEulerSuite = test_euler.suite()
loadMat33Suite = test_mat33.suite()
loadMat44Suite = test_mat44.suite()
loadQuatSuite = test_quat.suite()
loadRotationOrderSuite = test_rotation_order.suite()
loadXfoSuite = test_xfo.suite()


def suite():
    suites = [
        loadVec2Suite,
        loadVec3Suite,
        loadVec4Suite,
        loadColorSuite,
        loadEulerSuite,
        loadMat33Suite,
        loadMat44Suite,
        loadQuatSuite,
        loadRotationOrderSuite,
        loadXfoSuite]

    return unittest.TestSuite(suites)


if __name__ == '__main__':
    runner = unittest.TextTestRunner()
    runner.run(suite())
