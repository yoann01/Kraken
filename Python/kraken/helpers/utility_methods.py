from kraken.core.maths.math_object import MathObject
from kraken.core.maths.vec3 import Vec3
from kraken.core.maths.xfo import Xfo
from kraken.core.maths.quat import Quat

from kraken.core.maths import decodeValue


def logHierarchy(kObject):
    """Traverses the given Kraken hierarchy and logs the names of all the objects.

    Args:
        Object: The object to start logging on.

    """

    print kObject.getDecoratedPath()
    for i in xrange(kObject.getNumChildren()):
        child = kObject.getChildByIndex(i)
        logHierarchy(child)


def __convertFromJSON(jsonData):

    if type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertFromJSON(item))
        return newList
    elif type(jsonData) is dict:
        if '__mathObjectClass__' in jsonData.keys():
            return decodeValue(jsonData)
        for key, value in jsonData.iteritems():
            jsonData[key] = __convertFromJSON(value)
    return jsonData


def prepareToLoad(jsonData):
    """Prepares the json data for loading into kraken.

    Args:
        jsonData (dict): The JSON data to be prepared.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __convertFromJSON(jsonData)


def __convertToJSON(jsonData):

    if isinstance(jsonData, MathObject):
        return jsonData.jsonEncode()
    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__convertToJSON(item))
        return newList
    elif type(jsonData) is dict:
        newDict = {}
        for key, value in jsonData.iteritems():
            newDict[key] = __convertToJSON(value)
        return newDict
    return jsonData


def prepareToSave(jsonData):
    """Prepares the json data for serialization.

    Args:
        jsonData (dict): The JSON data to be prepared.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __convertToJSON(jsonData)


def __mirrorData(jsonData, plane):

    if isinstance(jsonData, Vec3):
        newVec3 = Vec3(jsonData)
        if plane == 0:
            newVec3.x = -newVec3.x
        elif plane == 1:
            newVec3.y = -newVec3.y
        elif plane == 2:
            newVec3.z = -newVec3.z
        return newVec3

    if isinstance(jsonData, Quat):
        newQuat = Quat(jsonData)
        newQuat.mirror(plane)
        return newQuat

    elif isinstance(jsonData, Xfo):
        newXfo = Xfo(jsonData)
        if plane == 0:
            newXfo.tr.x = -newXfo.tr.x
        elif plane == 1:
            newXfo.tr.y = -newXfo.tr.y
        elif plane == 2:
            newXfo.tr.z = -newXfo.tr.z

        newXfo.ori.mirror(plane)
        return newXfo

    elif type(jsonData) is list:
        newList = []
        for item in jsonData:
            newList.append(__mirrorData(item, plane))
        return newList

    elif type(jsonData) is dict:
        newDict = {}
        for key, value in jsonData.iteritems():
            newDict[key] = __mirrorData(value, plane)
        return newDict

    return jsonData


def mirrorData(jsonData, plane):
    """Prepares the json data for serialization.

    Args:
        jsonData (dict): The JSON data to be prepared.
        plane (int): The plane to mirror across.

    Returns:
        dict: The prepared JSON hierarchy.

    """

    return __mirrorData(jsonData, plane)
