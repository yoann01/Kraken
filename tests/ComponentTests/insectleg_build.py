from kraken import plugins
from kraken.core.maths import Vec3
from kraken_components.generic.insectleg_component import InsectLegComponentGuide, InsectLegComponentRig

from kraken.core.profiler import Profiler
from kraken.helpers.utility_methods import logHierarchy


Profiler.getInstance().push("insectLeg_build")

insectLegGuide = InsectLegComponentGuide("insectLeg")
insectLegGuide.loadData({
    "name": "insectLeg",
    "location": "L",
    "numJoints": 4,
    "jointPositions": [Vec3(0.9811, 9.769, -1.237),
                       Vec3(5.4488, 8.4418, -1.237),
                       Vec3(4.0, 3.1516, -1.237),
                       Vec3(6.841, 1.0, -1.237),
                       Vec3(9.841, 0.0, -1.237)]
})

# Save the hand guide data for persistence.
saveData = insectLegGuide.saveData()

insectLegGuideData = insectLegGuide.getRigBuildData()

leg = InsectLegComponentRig()
leg.loadData(insectLegGuideData)

builder = plugins.getBuilder()
builder.build(leg)

Profiler.getInstance().pop()

if __name__ == "__main__":
    print Profiler.getInstance().generateReport()
else:
    for each in leg.getItems().values():
        # Only log hierarchy for Layer objects as Layers in this test are added to
        # the component since there is no rig object.
        if each.isTypeOf('Layer'):
            logHierarchy(each)
