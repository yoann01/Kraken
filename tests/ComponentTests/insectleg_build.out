Failed to find DCC builder. Falling back to Python builder.
building: controls.insectLeg as: insectLeg_L_cmp type: ComponentGroup
buildGroup: controls.insectLeg as: insectLeg_L_cmp
building: controls.insectLeg.inputs as: insectLeg_L_inputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.insectLeg.inputs as: insectLeg_L_inputs_hrc
building: controls.insectLeg.inputs.rootInput as: insectLeg_L_rootInput_cmpIn type: ComponentInput
buildLocator: controls.insectLeg.inputs.rootInput as: insectLeg_L_rootInput_cmpIn
building: controls.insectLeg.outputs as: insectLeg_L_outputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.insectLeg.outputs as: insectLeg_L_outputs_hrc
building: controls.insectLeg.outputs.legEndXfoOutput as: insectLeg_L_legEndXfoOutput_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.legEndXfoOutput as: insectLeg_L_legEndXfoOutput_cmpOut
building: controls.insectLeg.outputs.legEndPosOutput as: insectLeg_L_legEndPosOutput_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.legEndPosOutput as: insectLeg_L_legEndPosOutput_cmpOut
building: controls.insectLeg.outputs.bone01 as: insectLeg_L_bone01_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.bone01 as: insectLeg_L_bone01_cmpOut
building: controls.insectLeg.outputs.bone02 as: insectLeg_L_bone02_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.bone02 as: insectLeg_L_bone02_cmpOut
building: controls.insectLeg.outputs.bone03 as: insectLeg_L_bone03_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.bone03 as: insectLeg_L_bone03_cmpOut
building: controls.insectLeg.outputs.bone04 as: insectLeg_L_bone04_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.bone04 as: insectLeg_L_bone04_cmpOut
building: controls.insectLeg.ChainBase as: insectLeg_L_ChainBase_loc type: Locator
buildLocator: controls.insectLeg.ChainBase as: insectLeg_L_ChainBase_loc
building: controls.insectLeg.bone01FK as: insectLeg_L_bone01FK_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.bone01FK as: insectLeg_L_bone01FK_ctrlSpace
building: controls.insectLeg.bone01FK.bone01FK as: insectLeg_L_bone01FK_ctrl type: Control
buildControl:controls.insectLeg.bone01FK.bone01FK as: insectLeg_L_bone01FK_ctrl
building: controls.insectLeg.bone01FK.bone01FK.bone02FK as: insectLeg_L_bone02FK_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.bone01FK.bone01FK.bone02FK as: insectLeg_L_bone02FK_ctrlSpace
building: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK as: insectLeg_L_bone02FK_ctrl type: Control
buildControl:controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK as: insectLeg_L_bone02FK_ctrl
building: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK as: insectLeg_L_bone03FK_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK as: insectLeg_L_bone03FK_ctrlSpace
building: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK as: insectLeg_L_bone03FK_ctrl type: Control
buildControl:controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK as: insectLeg_L_bone03FK_ctrl
building: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK.bone04FK as: insectLeg_L_bone04FK_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK.bone04FK as: insectLeg_L_bone04FK_ctrlSpace
building: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK.bone04FK.bone04FK as: insectLeg_L_bone04FK_ctrl type: Control
buildControl:controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK.bone04FK.bone04FK as: insectLeg_L_bone04FK_ctrl
building: controls.insectLeg.IK as: insectLeg_L_IK_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.IK as: insectLeg_L_IK_ctrlSpace
building: controls.insectLeg.IK.IK as: insectLeg_L_IK_ctrl type: Control
buildControl:controls.insectLeg.IK.IK as: insectLeg_L_IK_ctrl
building: controls.insectLeg.UpV as: insectLeg_L_UpV_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.UpV as: insectLeg_L_UpV_ctrlSpace
building: controls.insectLeg.UpV.UpV as: insectLeg_L_UpV_ctrl type: Control
buildControl:controls.insectLeg.UpV.UpV as: insectLeg_L_UpV_ctrl
building: deformers as: deformers type: Layer
buildLayer: deformers as: deformers
building: deformers.insectLeg as: insectLeg_L_cmp type: ComponentGroup
buildGroup: deformers.insectLeg as: insectLeg_L_cmp
building: deformers.insectLeg.bone01 as: insectLeg_L_bone01_def type: Joint
buildJoint: deformers.insectLeg.bone01 as: insectLeg_L_bone01_def
building: deformers.insectLeg.bone02 as: insectLeg_L_bone02_def type: Joint
buildJoint: deformers.insectLeg.bone02 as: insectLeg_L_bone02_def
building: deformers.insectLeg.bone03 as: insectLeg_L_bone03_def type: Joint
buildJoint: deformers.insectLeg.bone03 as: insectLeg_L_bone03_def
building: deformers.insectLeg.bone04 as: insectLeg_L_bone04_def type: Joint
buildJoint: deformers.insectLeg.bone04 as: insectLeg_L_bone04_def
building: controls as: controls type: Layer
buildLayer: controls as: controls
building: controls.insectLeg as: insectLeg_L_cmp type: ComponentGroup
buildGroup: controls.insectLeg as: insectLeg_L_cmp
building: controls.insectLeg.inputs as: insectLeg_L_inputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.insectLeg.inputs as: insectLeg_L_inputs_hrc
building: controls.insectLeg.inputs.rootInput as: insectLeg_L_rootInput_cmpIn type: ComponentInput
buildLocator: controls.insectLeg.inputs.rootInput as: insectLeg_L_rootInput_cmpIn
building: controls.insectLeg.outputs as: insectLeg_L_outputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.insectLeg.outputs as: insectLeg_L_outputs_hrc
building: controls.insectLeg.outputs.legEndXfoOutput as: insectLeg_L_legEndXfoOutput_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.legEndXfoOutput as: insectLeg_L_legEndXfoOutput_cmpOut
building: controls.insectLeg.outputs.legEndPosOutput as: insectLeg_L_legEndPosOutput_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.legEndPosOutput as: insectLeg_L_legEndPosOutput_cmpOut
building: controls.insectLeg.outputs.bone01 as: insectLeg_L_bone01_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.bone01 as: insectLeg_L_bone01_cmpOut
building: controls.insectLeg.outputs.bone02 as: insectLeg_L_bone02_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.bone02 as: insectLeg_L_bone02_cmpOut
building: controls.insectLeg.outputs.bone03 as: insectLeg_L_bone03_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.bone03 as: insectLeg_L_bone03_cmpOut
building: controls.insectLeg.outputs.bone04 as: insectLeg_L_bone04_cmpOut type: ComponentOutput
buildLocator: controls.insectLeg.outputs.bone04 as: insectLeg_L_bone04_cmpOut
building: controls.insectLeg.ChainBase as: insectLeg_L_ChainBase_loc type: Locator
buildLocator: controls.insectLeg.ChainBase as: insectLeg_L_ChainBase_loc
building: controls.insectLeg.bone01FK as: insectLeg_L_bone01FK_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.bone01FK as: insectLeg_L_bone01FK_ctrlSpace
building: controls.insectLeg.bone01FK.bone01FK as: insectLeg_L_bone01FK_ctrl type: Control
buildControl:controls.insectLeg.bone01FK.bone01FK as: insectLeg_L_bone01FK_ctrl
building: controls.insectLeg.bone01FK.bone01FK.bone02FK as: insectLeg_L_bone02FK_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.bone01FK.bone01FK.bone02FK as: insectLeg_L_bone02FK_ctrlSpace
building: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK as: insectLeg_L_bone02FK_ctrl type: Control
buildControl:controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK as: insectLeg_L_bone02FK_ctrl
building: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK as: insectLeg_L_bone03FK_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK as: insectLeg_L_bone03FK_ctrlSpace
building: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK as: insectLeg_L_bone03FK_ctrl type: Control
buildControl:controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK as: insectLeg_L_bone03FK_ctrl
building: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK.bone04FK as: insectLeg_L_bone04FK_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK.bone04FK as: insectLeg_L_bone04FK_ctrlSpace
building: controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK.bone04FK.bone04FK as: insectLeg_L_bone04FK_ctrl type: Control
buildControl:controls.insectLeg.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK.bone04FK.bone04FK as: insectLeg_L_bone04FK_ctrl
building: controls.insectLeg.IK as: insectLeg_L_IK_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.IK as: insectLeg_L_IK_ctrlSpace
building: controls.insectLeg.IK.IK as: insectLeg_L_IK_ctrl type: Control
buildControl:controls.insectLeg.IK.IK as: insectLeg_L_IK_ctrl
building: controls.insectLeg.UpV as: insectLeg_L_UpV_ctrlSpace type: CtrlSpace
buildGroup: controls.insectLeg.UpV as: insectLeg_L_UpV_ctrlSpace
building: controls.insectLeg.UpV.UpV as: insectLeg_L_UpV_ctrl type: Control
buildControl:controls.insectLeg.UpV.UpV as: insectLeg_L_UpV_ctrl
building: deformers.insectLeg as: insectLeg_L_cmp type: ComponentGroup
buildGroup: deformers.insectLeg as: insectLeg_L_cmp
building: deformers.insectLeg.bone01 as: insectLeg_L_bone01_def type: Joint
buildJoint: deformers.insectLeg.bone01 as: insectLeg_L_bone01_def
building: deformers.insectLeg.bone02 as: insectLeg_L_bone02_def type: Joint
buildJoint: deformers.insectLeg.bone02 as: insectLeg_L_bone02_def
building: deformers.insectLeg.bone03 as: insectLeg_L_bone03_def type: Joint
buildJoint: deformers.insectLeg.bone03 as: insectLeg_L_bone03_def
building: deformers.insectLeg.bone04 as: insectLeg_L_bone04_def type: Joint
buildJoint: deformers.insectLeg.bone04 as: insectLeg_L_bone04_def
deformers
deformers.insectLeg:L
deformers.insectLeg:L.bone01
deformers.insectLeg:L.bone02
deformers.insectLeg:L.bone03
deformers.insectLeg:L.bone04
controls
controls.insectLeg:L
controls.insectLeg:L.inputs
controls.insectLeg:L.inputs.rootInput
controls.insectLeg:L.outputs
controls.insectLeg:L.outputs.legEndXfoOutput
controls.insectLeg:L.outputs.legEndPosOutput
controls.insectLeg:L.outputs.bone01
controls.insectLeg:L.outputs.bone02
controls.insectLeg:L.outputs.bone03
controls.insectLeg:L.outputs.bone04
controls.insectLeg:L.ChainBase
controls.insectLeg:L.bone01FK
controls.insectLeg:L.bone01FK.bone01FK
controls.insectLeg:L.bone01FK.bone01FK.bone02FK
controls.insectLeg:L.bone01FK.bone01FK.bone02FK.bone02FK
controls.insectLeg:L.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK
controls.insectLeg:L.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK
controls.insectLeg:L.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK.bone04FK
controls.insectLeg:L.bone01FK.bone01FK.bone02FK.bone02FK.bone03FK.bone03FK.bone04FK.bone04FK
controls.insectLeg:L.IK
controls.insectLeg:L.IK.IK
controls.insectLeg:L.UpV
controls.insectLeg:L.UpV.UpV
